### PR TITLE
Constrain cost function

### PR DESCRIPTION
**Pull request by oliverlee**

### CODE REVIEW

File: src/concepts.hpp [hunk 1]
```diff
@@ -120,6 +120,22 @@ concept Point =
...
   Subtractible<const T&, const U&, T>;
 
+template <Arithmetic, std::size_t>
+struct point;
+
+template <Arithmetic>
+struct dual;
+
+// TODO replace total ordering requirement with partial ordering?
+// TODO require Point to be rebindable? Otherwise we rebind into `opt::point`
```
Comment (Aertalon) [inline src/concepts.hpp:130]: Yes, we need to be able to always create a dual from a Point in the "natural" way (if I understand correctly)

File: src/concepts.hpp [hunk 2]
```diff
@@ -120,6 +120,22 @@ concept Point =
...
+//      for the gradient calculation
+template <class T, class P>
+concept Cost =
```
Comment (Aertalon) [inline src/concepts.hpp:133]: Nice. We can build on top of this a ConvexCost, hopefully

File: src/convopt.hpp [hunk 3]
```diff
@@ -13,7 +13,7 @@ namespace opt {
 // TODO(enrlov): F must be invocable with P and opt::point<dual<scalar_t<P>>>
```
Comment (Aertalon) [inline src/convopt.hpp:13]: To be removed

File: src/spaces.hpp [hunk 4]
```diff
@@ -14,6 +14,39 @@
...
+    // TODO move to common location or replace with constexpr math lib
+    constexpr auto abs = [](auto x) {
+        if (x < decltype(x){}) {
+            return -x;
+        }
+
+        return x;
+    };
+
+    return
```
Comment (Aertalon) [inline src/spaces.hpp:31]: So close_to now returns the evaluation of a lambda? Neat.

File: src/concepts.hpp [hunk 5]
```diff
@@ -120,6 +120,22 @@ concept Point =
   Subtractible<const T&, const T&, U> &&
   Subtractible<const T&, const U&, T>;
 
+template <Arithmetic, std::size_t>
+struct point;
+
+template <Arithmetic>
+struct dual;
+
+// TODO replace total ordering requirement with partial ordering?
```
[the remark this answers is not in the document]
Comment (oliverlee) [reply src/concepts.hpp:129]: This needs to implement exposition-only _boolean-testable_ so I suggest we just do it later.

**Review by Aertalon — APPROVED** [review]

**Pull request merged**